### PR TITLE
Release/2.1.0

### PR DESCRIPTION
**Pull request by cs-yash-patodkar**
Release 2.1.0
Fixes :
- Increased width for "Top 3 playbooks" box
Addition
- Added a new label, "By automation" it shows alerts closed by automation (Key highlight)
- New filter in edit view : "Exclude playbooks with tags", to exclude playbooks in "Top 3 playbooks" with the mentioned tags
- Changed "Top 3 data sources" to "Top 3 detections"
- Can add data using JSON single record
UTC :
Installed widget successfully
Top 3 playbooks with "Exclude Tags" filter
By Automation number
Top 3 playbooks box's length

### CODE REVIEW

File: widget/edit.controller.js [hunk 1]
```diff
@@ -34,12 +70,43 @@
           $scope.params.fieldsArray = entity.getFormFieldsArray();
         });
       }
-      
+
+      function onChangeModuleType() {
+        delete $scope.config.query;
+        delete $scope.config.customModuleField;
```
Comment (cs-suyog-jinturkar) [inline widget/edit.controller.js:76]: You should have a reset function to delete or clear data

Comment (cs-yash-patodkar) [reply widget/edit.controller.js:76]: That function just resets 3 fields, I don't feel the need for more abstraction

File: widget/assets/socWidgetInput.json [hunk 2]
```diff
@@ -1,82 +1,218 @@
...
     },
-    "closed":{
-        "title":"Closed"
+    "closed": {
+        "title": "Closed"
     },
-    "topIncidents":{
-        "title":"Top 3 Incident Types"
+    "automatedResolved": {
+        "title": "by Automation"
```
Comment (cs-pranali-chavan) [inline widget/assets/socWidgetInput.json:31]: IMO title should be capitalise

File: widget/view.controller.js [hunk 3]
```diff
@@ -345,6 +457,21 @@
             'operator': 'count'
           }]
       };
+      if ($scope.config.recordTags) {
+        var playbookExcludeTag = []
+        for (var i = 0; i < $scope.config.recordTags.length; i++) {
+          var tagArr = $scope.config.recordTags[i].split('/');
```
Comment (cs-pranali-chavan) [inline widget/view.controller.js:463]: '/' will backslash be present in all recordTags or else a check can be applied

File: widget/edit.controller.js [hunk 4]
```diff
@@ -34,12 +70,43 @@
...
+        }),
+      appModulesService.load(true).then(function (modules) {
+        $scope.modules = modules;
+
+        //Create a list of modules with atleast one JSON field
+        modules.forEach((module, index) =>{
+          var moduleMetaData = modelMetadatasService.getMetadataByModuleType(module.type);
+          for(let fieldIndex =0; fieldIndex < moduleMetaData.attributes.length; fieldIndex++){
+            //Check If JSON field is present in the module
+            if(moduleMetaData.attributes[fieldIndex].type === "object"){
```
Comment (cs-pranali-chavan) [inline widget/edit.controller.js:99]: Also check if type object is not empty

**Review by cs-pranali-chavan — CHANGES_REQUESTED** [review]
Kindly check the comments. Rest of the changes are appropriate

File: widget/edit.controller.js (hunk 4 above)
Comment (cs-yash-patodkar) [reply widget/edit.controller.js:99]: Object does not have data, it just represents if there is a fields with type "json" in the module
Not fetching the value, rather the metadata of the module

File: widget/assets/socWidgetInput.json (hunk 2 above)
Comment (cs-yash-patodkar) [reply widget/assets/socWidgetInput.json:31]: It was discussed with Abhishek. No need to capitalise "by"

File: widget/view.controller.js (hunk 3 above)
Comment (cs-yash-patodkar) [reply widget/view.controller.js:463]: It has already been handled, if the string contains no '/' later pop is used which returns the last value in the array

**Review by cs-pranali-chavan — APPROVED** [review]
Comment response are valid.

**Pull request merged**